src/ray/rpc/raylet/raylet_client_pool.cc

```diff
@@ -96,6 +96,8 @@ std::shared_ptr<ray::RayletClientInterface> RayletClientPool::GetOrConnectByAddr
   return connection;
 }
 
+// NOTE: On node death the raylet client is removed from the pool. Hence, you should not
+// add a RAY_CHECK after you call GetByID for nullptr.
 std::shared_ptr<ray::RayletClientInterface> RayletClientPool::GetByID(ray::NodeID id) {
```

Comment:
```suggestion
std::shared_ptr<ray::RayletClientInterface> RayletClientPool::GetByAliveNodeID(ray::NodeID id) {
```
don't add the cryptic note, put it in the header comment for the method instead (dead nodes are removed)

Reply:
This is also an abstraction leak -- the raylet client pool doesn't know anything about node liveness. the semantics you're talking about are implemented by the upstream components